### PR TITLE
Add additional flamegraph tests

### PR DESCRIPTION
**Pull request by jasonrhansen**
Includes tests for the following from #63:
- Use a palette file
- Ignore bad lines
- Plot an empty trace
- Plot a file where frames get pruned
- Plot an inverted flame graph
- Include grey frames
- Test a non-sorted multi-reader input

### CODE REVIEW

File: tests/flamegraph.rs [hunk 1]
```diff
@@ -1,34 +1,65 @@
...
+    options: Options,
+) -> quick_xml::Result<()> {
     let r = File::open(input_file).unwrap();
     let expected_len = fs::metadata(expected_result_file).unwrap().len() as usize;
     let mut result = Cursor::new(Vec::with_capacity(expected_len));
-    flamegraph::from_reader(options, r, &mut result).unwrap();
-    let mut expected = BufReader::new(File::open(expected_result_file).unwrap());
+    let return_value = flamegraph::from_reader(options, r, &mut result);
+    let expected = BufReader::new(File::open(expected_result_file).unwrap());
+    result.set_position(0);
```
Comment (jonhoo) [inline tests/flamegraph.rs]: I don't think you explicitly have to set the position to 0?

File: tests/flamegraph.rs [hunk 2]
```diff
@@ -116,9 +167,128 @@ fn flamegraph_nameattr() {
...
+
+#[test]
+fn flamegraph_should_warn_about_fractional_samples() {
+    test_flamegraph_logs(
+        "./tests/data/fractional-samples/fractional.txt",
+        |captured_logs| {
+            assert_eq!(captured_logs.len(), 1);
+            assert!(captured_logs[0].body.contains("fractional sample"));
+            assert_eq!(captured_logs[0].level, Level::Warn);
+        },
```
Comment (jonhoo) [inline tests/flamegraph.rs]: We may want to make this test a _little_ more liberal in that it should test that there is exactly one occurrence of that warning (as opposed to _also_ requiring that there are no other outputs).

File: tests/flamegraph.rs [hunk 3]
```diff
@@ -116,9 +167,128 @@ fn flamegraph_nameattr() {
...
+    );
+}
+
+#[test]
+fn flamegraph_should_not_warn_about_zero_fractional_samples() {
+    test_flamegraph_logs(
+        "./tests/data/fractional-samples/zero-fractionals.txt",
+        |captured_logs| {
+            assert!(captured_logs.is_empty());
+        },
```
Comment (jonhoo) [inline tests/flamegraph.rs]: Here too, I think we probably want to be a little more liberal. Specifically, check that there are no messages about fractional numbers.

File: tests/flamegraph.rs [hunk 4]
```diff
@@ -116,9 +167,128 @@ fn flamegraph_nameattr() {
...
+
+    test_flamegraph(input_file, expected_result_file, options).unwrap();
+}
+
+#[test]
+fn flamegraph_should_warn_about_bad_input_lines() {
+    test_flamegraph_logs("./tests/data/bad-lines/bad-lines.txt", |captured_logs| {
+        assert_eq!(captured_logs.len(), 1);
+        assert!(captured_logs[0].body.contains("invalid format"));
+        assert_eq!(captured_logs[0].level, Level::Warn);
```
Comment (jonhoo) [inline tests/flamegraph.rs]: Don't we also print something along the lines of "N lines ignored"? I guess that's only in collapsing.
In any case, I think we should be liberal here too.

File: tests/flamegraph.rs [hunk 5]
```diff
@@ -116,9 +167,128 @@ fn flamegraph_nameattr() {
...
+    });
+}
+
+#[test]
+fn flamegraph_should_warn_about_empty_input() {
+    test_flamegraph_logs("./tests/data/empty/empty.txt", |captured_logs| {
+        assert_eq!(captured_logs.len(), 1);
+        assert_eq!(captured_logs[0].body, "No stack counts found");
+        assert_eq!(captured_logs[0].level, Level::Error);
```
Comment (jonhoo) [inline tests/flamegraph.rs]: Same here.

**Comment by jonhoo** [comment]
Looks good overall! Thanks for putting in the work :D
Left a few inline comments but they should be pretty minor!

File: tests/flamegraph.rs (hunk 1 above)
Comment (jasonrhansen) [reply tests/flamegraph.rs]: We're using `result` as both a writer and reader. It gets passed into `flamegraph::from_reader` to write the result into, which advances the cursor. Then when we're comparing we need to reset the cursor so we start reading from the beginning.

**Pull request merged**